### ISSUE TITLE
Support PMTiles in Editor

### ISSUE THREAD

**Issue by ohrie** [maputnik/editor#807]
PMTiles from [Protomaps](https://protomaps.com/) are increasingly popular and a lot of other services are [supporting](https://github.com/felt/tippecanoe/releases/tag/2.17.0) it now. It has some major advantages in comparison to MBTiles. Styling the PMTiles are, however not possible yet, using the data set directly. 

MapLibre GL JS [supports PMTiles](https://protomaps.com/docs/frontends/maplibre). Therefore a switch to Maplibre GL JS in addition with the support of pmtiles URLs link ([like this](https://protomaps.github.io/PMTiles/?url=https%3A%2F%2Fprotomaps.github.io%2FPMTiles%2Fprotomaps%28vector%29ODbL_firenze.pmtiles#map=12.56/43.7798/11.24148))  would be super cool, so we can use PMTiles in Maputnik!

**Comment by bogind** [comment]
MapLibre supports this feature with an external plugin that relies on the [addprotocol](https://maplibre.org/maplibre-gl-js/docs/API/classes/default/#addprotocol) method.  

![image](https://github.com/maputnik/editor/assets/26594971/dcca142d-279f-4c98-8cde-0faaf646d9cf)

I think that adding this functionality might confuse users who do not know this,  
and that may not know that the output of their style will only work under certain tile deployment and development conditions (for example it will likely not work in GeoServer or QGIS)

In my opinion, unless PMTiles are integrated as an official source in the [MapLibre Style Sepcification](https://maplibre.org/maplibre-style-spec/) this will likely cause confusion, even though it might be highly beneficial for certain users.

**Comment by lseelenbinder** [comment]
I agree with @bogind, I think it makes more sense to support this as a XYZ URL in Maputnik. It is very difficult to support plugins at the editor level, and it's not a significant workflow difference if you have to substitute one URL.

If and when it's integrated, then we can reconsider.

**Comment by zstadler** [comment]
Now that MapLibre is the rendering engine for Maputnik #812, it would make sense to me that the PMTiles plugin will be added by default to Maputnik, and enable the use of `pmtiles:...` sources.

**Comment by bogind** [comment]
> Now that MapLibre is the rendering engine for Maputnik #812, it would make sense to me that the PMTiles plugin will be added by default to Maputnik, and enable the use of `pmtiles:...` sources.

I don't think the issue was with the rendering engine, but with the style specification itself.
PMTiles is still an external plugin for MapLibre.

In my opinion (currently as a user, not a contributor), this should not be added, unless with strict instructions and warnings about compatibility. This can cause credibility issues with non-proficient users who would try to use the exported `pmtiles:...` styles in rendering engines that do not support them.

**Comment by ThomasG77** [comment]
To bypass the issue, host locally Maputnik or serve pmtiles as x/y/z.mvt

## Solution to use locally pmtiles with Maputnik

```bash
git clone https://github.com/maplibre/maputnik.git
cd maputnik
npm i
npm i pmtiles

echo -e "import { Protocol } from 'pmtiles';\n$(cat src/components/MapMaplibreGl.tsx)" >| src/components/MapMaplibreGl.tsx

sed -i 's#const map = new MapLibreGl.Map(mapOpts);#const map = new MapLibreGl.Map(mapOpts);\nlet protocol = new Protocol();\nMapLibreGl.addProtocol("pmtiles",protocol.tile);#g' src/components/MapMaplibreGl.tsx

mv dist maputnik
npm i -g http-server
http-server
```

Then open http://127.0.0.1:8080/maputnik/ and you can try using maputnik with pmtiles as source


## Alternative if you have a public server where you can install stuff with your pmtiles

No need to change anything on Maputnik public side. Use simply https://maputnik.github.io/editor/ and consume `{z}/{x}/{y}.mvt` URLs

```bash
# If another OS, see https://github.com/protomaps/go-pmtiles/releases
wget https://github.com/protomaps/go-pmtiles/releases/download/v1.19.2/go-pmtiles_1.19.2_Linux_x86_64.tar.gz
tar -xvzf go-pmtiles_1.19.2_Linux_x86_64.tar.gz
chmod +x pmtiles
# Sample file. Use you own here
wget https://r2-public.protomaps.com/protomaps-sample-datasets/cb_2018_us_zcta510_500k.pmtiles
./pmtiles serve --cors=* cb_2018_us_zcta510_500k.pmtiles
```

Change now in the Maplibre JSON style the block starting with `"pmtiles://..."` to `http://yourhost.domain:8081/cb_2018_us_zcta510_500k.pmtiles/{z}/{x}/{y}.mvt`, do your styling then you can switch back to `"pmtiles://..."` when in production.

**Comment by HarelM** [comment]
I believe MBtiles and PMtiles are basically the same, you can't use them as is in maplibre, therefore you either need to serve them, even from local machine using a tile server, or host them somewhere.
There are great instructions that I have followed in order to serve a pmtiles file from cloud flare.
You can use martin for local development, or tileserver-gl.
All of those solutions will allow you to create a style that can easily work on any maplibre client.
I'm not sure adding the pmtiles plugin and supporting its protocol will benefit the users.
I think that a readme with a section about hosting local tile files such as PMtiles and MBtiles can solve most users' problems.

**Comment by diehl** [comment]
@HarelM Would you mind sharing the instructions for serving PMTiles from CloudFlare? I stumbled on this post while searching for information on designing a custom basemap style with PMTiles as the underlying data source. Sounds like you just provided me the pathway to a solution!

**Comment by HarelM** [comment]
Pm tiles has excellent instructions:
https://docs.protomaps.com/pmtiles/cloud-storage

**Comment by diehl** [comment]
Thank you @HarelM!

**Comment by josxha** [comment]
In the meantime Protomaps has a hosted version of Maputnik with support for PMTiles.
https://editor.protomaps.com/
This probably doesn't really solves this issue since it's a forked of Maputnik but should be helpful for everyone who wants to edit a protomaps compatible style. Only thing to care about is that the version of Maputnik hosted on Protomaps is slightly outdated.

**Comment by HarelM** [comment]
According to what I've seen there, you still need to host the PMTiles file locally.
For hosting a local file you can use a small tile server like Martin, and then you can use maputnik as is...

**Comment by bdon** [comment]
> According to what I've seen there, you still need to host the PMTiles file locally.

http://editor.protomaps.com works with remote PMTiles archives, there's no limitation on localhost or local files etc. I have updated the fork with the most recent changes from maputnik/main.

**Comment by prusswan** [comment]
> > According to what I've seen there, you still need to host the PMTiles file locally.
> 
> http://editor.protomaps.com works with remote PMTiles archives, there's no limitation on localhost or local files etc. I have updated the fork with the most recent changes from maputnik/main.

I tried adding the inspect behavior used in the editor to the tiles viewer (pmtiles.io) just to see how it works with local files. Seems like something is still missing (internal behavior like `map.style.sourceCaches[sourceId]._source.vectorLayerIds`?). Maybe it will be easier to add drag n drop support to the editor instead (should work well for both small pmtiles packages and style.json)

**Comment by bdon** [comment]
Yeah this is a bug right now internally in the PMTiles MapLibre adapter with how it deals with fetching TileJSON for local archives.

For Maputnik there is no UI element for drag-and-drop, and the user experience may not be ideal since it requires user interaction to open a local file. For the hosted maputnik you could communicate with localhost using a self-signed SSL certificate.

**Comment by prusswan** [comment]
> Yeah this is a bug right now internally in the PMTiles MapLibre adapter with how it deals with fetching TileJSON for local file archives.

Turns out that you just need a way to pass `metadata.vector_layers` to `map.style.sourceCaches["source"]._source.vectorLayerIds`. I just store it under `style["metadata"]` but this hack is good as any.

The internal API that maplibre-gl-inspect is relying on: https://github.com/maplibre/maplibre-gl-inspect/blob/5e234935bba74e2870dae0df57098c48bc3be9ef/lib/MaplibreInspect.ts#L213

> For Maputnik there is no UI element for drag-and-drop, and the user experience may not be ideal since it requires user interaction to open a local file. For the hosted maputnik you could communicate with localhost using a self-signed SSL certificate.

I guess you mean it is not easy to find a good spot for the UI element, not that it can't be done? It would be nice to be able to drop a local file onto https://editor.protomaps.com with the style pre-populated according to the metadata, probably the minimal interaction needed for those who prefer this option.

**Comment by josxha** [comment]
Support for vector PMTiles has been added in https://github.com/maplibre/maputnik/pull/938. Does this issue track raster tiles as well or can it be closed?

**Comment by miiimooo** [comment]
@josxha are you sure? I've tested this but I get an error in the browser console:
> MapMaplibreGl.tsx:124 Fetch API cannot load pmtiles://https//demo-bucket.protomaps.com/v4.pmtiles. URL scheme "pmtiles" is not supported.

To reproduce:
* Data Sources
* Add New Source
* Source Type: Vector (PMTiles)
* PMTiles URL: https://demo-bucket.protomaps.com/v4.pmtiles

**Comment by josxha** [comment]
@miiimooo I can see these errors in the console

```console
maplibre-gl-inspect.mjs:2850 
 Fetch API cannot load pmtiles://https//demo-bucket.protomaps.com/v4.pmtiles. URL scheme "pmtiles" is not supported.
maplibre-gl-inspect.mjs:2861 
 Unable to retrieve tileJSON from pmtiles://https://demo-bucket.protomaps.com/v4.pmtiles using style's layers
maplibre-gl-inspect.mjs:2850 
 Fetch API cannot load pmtiles://https//demo-bucket.protomaps.com/v4.pmtiles. URL scheme "pmtiles" is not supported.
maplibre-gl-inspect.mjs:2861 
 Unable to retrieve tileJSON from pmtiles://https://demo-bucket.protomaps.com/v4.pmtiles using style's layers
```


Though, tiles still load fine for me:

<img width="2981" height="1737" alt="Image" src="https://github.com/user-attachments/assets/67c9847b-2837-444e-a9dc-4b166263eaa9" />

**Comment by miiimooo** [comment]
@josxha so it's loading something but not displaying anything?

**Comment by josxha** [comment]
@miiimooo It's loading and displaying something.

https://github.com/user-attachments/assets/26ec7b96-94a3-4da3-9e95-f538679f4967

**Comment by bdon** [comment]
The console errors are normal and come from the `maplibre-gl-inspect` plugin trying multiple ways to fetch TileJSON, they don't indicate the `.pmtiles` is failing to load.

**Comment by jo-chemla** [comment]
On live/prod [https://maplibre.org/maputnik](https://maplibre.org/maputnik/) I'm also seeing below error trying to open the [geotrek.json](https://ignf.github.io/cartes-ign-app/geotrek.json) style from french IGN - either opening from remote json endpoint url or copy-pasting json content into the `Code editor`. 

You can have a look at the underlying pmtiles dataset on [pmtiles viewer](https://pmtiles.io/#url=https://data.geopf.fr/chunk/telechargement/download/HIKING.GEOTREK/archive/data.pmtiles&map=5.8/46.473/4.955)

> Fetch API cannot load pmtiles://https//data.geopf.fr/chunk/telechargement/download/HIKING.GEOTREK/archive/data.pmtiles. URL scheme "pmtiles" is not supported.


EDIT: Discard that comment, it's actually the `layer > layout > "icon-image": "pill"` which was failing maplibre since it might have no knowledge of such an icon in the sprite set. The error was thrown by `maplibre-gl-inspect.mjs` as stated above

<details><summary>Working geotrek style</summary>
<p>

```JSON

A Maputnik safe version of the IGN geotek style: 
{
  "version": 8,
  "name": "Idées randos (Maputnik-safe)",
  "glyphs": "https://data.geopf.fr/annexes/ressources/vectorTiles/fonts/{fontstack}/{range}.pbf",
  "sprite": "https://ignf.github.io/cartes-ign-app/data/poi-osm-sprite",
  "sources": {
    "geotrek": {
      "type": "vector",
      "url": "pmtiles://https://data.geopf.fr/chunk/telechargement/download/HIKING.GEOTREK/archive/data.pmtiles"
    }
  },
  "layers": [
    { "id": "bg", "type": "background", "paint": { "background-color": "#f5f5f2" } },
    {
      "id": "geotrek-dot", "type": "circle", "source": "geotrek", "source-layer": "geotrek",
      "paint": {
        "circle-radius": ["interpolate",["linear"],["zoom"],6,2,14,6],
        "circle-color": ["match",["get","pratique_norm"],
          "Pédestre","#d13b3b","Cyclo","#2b7bba","Équestre","#8a5a2b","#666"],
        "circle-stroke-color":"#fff","circle-stroke-width":1
      }
    },
    {
      "id": "geotrek-label", "type": "symbol", "source": "geotrek", "source-layer": "geotrek",
      "minzoom": 10,
      "layout": {
        "text-field": ["concat",["get","nom_itineraire"],"  (",["to-string",["get","kilometers"]]," km)"],
        "text-font": ["Source Sans Pro Semibold"], "text-size": 12,
        "text-offset": [0,1.1], "text-anchor": "top", "text-optional": true
      },
      "paint": { "text-color":"#222","text-halo-color":"#fff","text-halo-width":1.4 }
    }
  ]
}

```

</p>
</details>

**Comment by HarelM** [comment]
I've recently updated to maplibre v6, might be related to the upgrade. 
Feel free to look into it.
